### ISSUE TITLE
copy all / paste all

### ISSUE THREAD

**Issue by TurboGit** [darktable-org/darktable#3153]
**Describe the bug**
copy all / paste all from lighttable completely destroy the edits

**To Reproduce**
1. Select one image
2. Got to darkroom
3. Activate module monochrome (easy effect to see)
4. Got back to lighttable
5. Select modified image, do copy all (or ctrl + c)
6. Select all images, do paste all (or ctrl + v)
7. Many images are fully black, open it in darkroom, the history stack is fully broken

**Expected behavior**
Only monochrome is added on all image history

**Platform (please complete the following information):**
 - OS Debian
 - Version 2.7.0+2009~g896cbe3c4-dirty

**Comment by jenshannoschwalm** [comment]
For me ^C ^V works fine if i have **all** images having **same** iop_order.

**Comment by jenshannoschwalm** [comment]
Something not directly related; is there a list what "things" might be / are delegated to background jobs?

**Comment by TurboGit** [comment]
I forgot one step to reproduce sorry:

0. Select all images and reset history stack

**Comment by TurboGit** [comment]
No I've not seen such a list.

**Comment by jenshannoschwalm** [comment]
One more question:
Is it even possible to add **all** modules? There are some with clearly specific parameters, basecurve, lens, rawblack, inputprofile and maybe more. I don't know atm. Any clear answers on that?

**Comment by junkyardsparkle** [comment]
The basecurve copies and pastes it actual settings, even when copied from an image where it was "auto-applied" by having that option checked in preferences (just tried it). The lens correction module OTOH will only copy and paste this way if it's been manually changed from the auto-detected settings, otherwise simply activates auto-detect on pasted-to image. Possibly other modules with autodetect (profiled denoise, input profile) do the same, so probably each would have to be checked.

I also noticed just now that if I discard history for a selected image, then us Ctrl+Shift+C to try to bring up the selective copy dialog, I get flashed a message that images without history can't be copied... but if I just use Ctrl+C, there is no such warning... odd.

**Comment by TurboGit** [comment]
@jenshannoschwalm : I would say that it is possible to add all modules, but the result would probably be desastrous at the user's level :)

I think the issue here is that we paste to image with iop-version-order = 0 and so at some point the image will be migrated to v5 but with some history. Probably quite wrong indeed. We may just need to migrate image to latest iop-order-version, adding default modules before pasting the new history.

**Comment by jenshannoschwalm** [comment]
>  but the result would probably be desastrous at the user's level :)

Right. But i don't care as long as dt keep running in a stable state :-) I am really concerned about rawblack. We have hidden options working on sensor data. Should we exclude this from merging?

> We may just need to migrate image to latest iop-order-version, adding default modules before pasting the new history.

I'll do that.

**Comment by TurboGit** [comment]
@jenshannoschwalm : I think default module are added at import time, so we may only need to set the image's iop-order-version.

**Comment by jenshannoschwalm** [comment]
> I think the issue here is that we paste to image with iop-version-order = 0 and so at some point the image will be migrated to v5 but with some history. Probably quite wrong indeed. We may just need to migrate image to latest iop-order-version, adding default modules before pasting the new history.

It doesn't work this way.

It is slightly more tricky than i thought initially but now i am sure i understand the two problems related to copy&paste history.

1. We must change the iop_order of each module if the iop_order version is not the same. No problem here.
2. We introduced the inclusion/auto-apply of modules to history a while ago. This implies, if we paste the complete history we also insert modules to the history which were **not** inserted in previous versions (before exposing **all** modules in history).

Technically an **exact replacement of** the history or **exact adding to** the history are both simply "wrong" (image specific parameters like colorin, base curve, sensor specifics ...) **Also** this is **not** what we had before and probably not what a user would expect.

So lets ask _"What is the desired behaviour of pasting a history?"_

I guess it means: _"add steps you did while developing image A to another image."_

So how to implement? I opt for: "Make sure (for the target image) you have a proper history with auto-applied modules and after that copy **only** items from the source history that are not auto-applied"
@TurboGit and @parafin do you agree on this?

We also have the history-selecting option, keep it as is?

**Comment by parafin** [comment]
Settings for auto-applied modules can be changed too... To me copy all + paste all looks almost useless. Consider changing white balance in original picture, then compressing the history (so that there's only one white balance iop in the history) and then doing copy/paste all. Should white balance settings be copied or not? How to distinguish that case from not changed white balance?

**Comment by TurboGit** [comment]
I think we are speaking about the always ON auto-applied modules. I think it makes sense to copy only the modules that are not ALWAYS ON as it was the case in previous version (where we did not exposed the auto-applied modules).

In 2.6.2 the copy was listing white-balance module only if the params where not the default and this is still the case with current master. This part if ok to me.

So the now listed modules in history:
* raw black/while point
* highlight reconstruction
* demosaic
* input color profile
* output color profile
* gamma 
should not be copied (in copy all) and not listed in copy dialog.

So to recap:
- copy all should NOT copy new exposed modules listed above.
- copy part should NOT list the new exposed modules listed above.

**Comment by parafin** [comment]
So, as an example, there would be no way to copy a setting for different highlight reconstruction algorithm?

**Comment by jenshannoschwalm** [comment]
I completely agree on copy-all. Not so sure about copy part. I use some of those mentioned if i know exactly what i do :-)

**Comment by TurboGit** [comment]
Sorry I probably meant that they should not be part of the copy if the params are default as for 2.6.2. That means all of them can be listed IINM expect gamma. Right?

**Comment by parafin** [comment]
So the check could go something like that:
```
if(!(module->default_enabled && !memcmp(hist->params, module->default_params, module->params_size))
```
I guess that could work.
P.S.
gamma should probably get IOP_FLAGS_NO_HISTORY_STACK, I don't see the point in having it in the history... Does it do anything at all BTW?

**Comment by TurboGit** [comment]
@parafin : yes, looks like something like that should be ok.

For gamma we can't remove it from history as this is a fence used in different cases and must be at the proper iop-order I think. And for example in pixelpipe_hb.c there is check for gamma as fence.

**Comment by parafin** [comment]
I don't quite understand why hidden module should be in the history. It won't disappear from pipe. Take finalscale iop as an example.

**Comment by parafin** [comment]
gamma is always the last iop and it's hidden, so it can be moved anywhere. Why anything will break if it's removed from history?

And on the other hand if you are right, then IOP_FLAGS_NO_HISTORY_STACK should be removed from finalscale too.

**Comment by TurboGit** [comment]
> gamma is always the last iop and it's hidden, so it can't be moved anywhere. Why anything will break if it's removed from history?

To not have a breakage we will need to take gamma as a special case in many places to ensure its iop-order is always last in the list. This means lot more code when pasting history, reading xmp... I'm not sure we are ready for that and I'm not 100% clear about all this business actually. And your point is interesting, what should we do for finalscale ? I'm don't know what it is used for.

That being said, for 3.0 we can't create an earthquake in the code at this stage.

**Comment by jenshannoschwalm** [comment]
imho for 3.0 we can only make iop_orders work _stable_ right now.

There are lots of tests & checks in iop_order code that might or might not be ok. We need better checks and rules. We must discuss such rules in detail before implementing.

**Comment by parafin** [comment]
I agree that if there's a suspicion that adding IOP_FLAGS_NO_HISTORY_STACK can break stuff, then it shouldn't be done for 3.0. But then it is probably already broken for finalscale, overexposed and rawoverexposed, that needs to be tested or/and thought through. finalscale is used for HQ export option, when scaling happens late in the pipe, other two as self-explanatory.

**Comment by TurboGit** [comment]
Fixed now.
